autoscheduler/scraper/management/commands/utils/pdf_parser.py

```diff
@@ -0,0 +1,145 @@
+""" (Modified) From Gabriel Britain's Good Bull Scheduler """
+
+from typing import Dict, List
+import re
+from dataclasses import dataclass
+
+import PyPDF2
+
+from scraper.management.commands.utils.pdf_helper import get_pdf_skip_count
+
+LEN_SECTION_ROW = 20
+LETTERS = ["A", "B", "C", "D", "F", "I", "S", "U", "Q", "X"]
+
+@dataclass # This automatically adds a constructor so we don't have to
+class GradeData():
+    """ The collection of grades for a particular section """
+    dept: str
+    course_num: str
+    section_num: str
+    letter_grades: Dict
+    gpa: float
+
+def sanitize_page(page_obj: PyPDF2.pdf.PageObject) -> List[str]:
+    """ Splits a PageObject's content on any number of newlines, and returns
+        the content as a list of strings.
+
+        Args:
+            page_obj: The content of a PDF page
+        Returns:
+            A list of strings representing the content on the page.
+    """
+    text = page_obj.extractText()
+    text = re.split(r"\n+", text)
+    return [t.strip() for t in text]
+
+def extract_letter_grades(section_row: List[str], old_pdf_style: bool) -> Dict[str, int]:
+    """ Maps the according section_row text to a dict w/ the respective letter grades &
+        how many people received that grade
+
+        Args:
+            section_row: The current text for this row
+            old_pdf_style: Whether this is the old pdf style (pre-2017)
+        Returns:
+            A dictionary of letter grades, with the key as the letter and the count of how
+            many people received that letter grade
+    """
+
+    abcdf_slice = slice(1, 10, 2)
+    isuqx_slice = slice(13, 18)
+    if old_pdf_style: # Old pdfs are arranged differently
+        abcdf_slice = slice(4, 9)
+        isuqx_slice = slice(10, 15)
+
+    letter_grades = section_row[abcdf_slice] + section_row[isuqx_slice]
+    letter_grades = {
+        l: int(grade) # Map each letter to the respective amount of grad
+        for l, grade in zip(LETTERS, letter_grades)
+    }
+
+    return letter_grades
+
+
+def parse_page(page_obj: PyPDF2.pdf.PageObject) -> List[GradeData]:
+    """ Parses a single page of a PDF, extracting a list of grade data for each section.
+
+        Args:
+            page_obj: A PyPDF2.pdf.PageObject representing the current page
+        Returns:
+            A list of GradeData objects, with the GPA as None
+    """
+
+    text = sanitize_page(page_obj) # Splits the text into separate lines
+    i = 0 # The current row in the pdf we're on
+    grade_data = [] # list of GradeData objects
+
+    while i < len(text):
+        skip_count = get_pdf_skip_count(text[i])
+        old_pdf_style = skip_count[0]
+        count = skip_count[1] # The actual count of how many rows should be skipped

```

Comment:
I don't know how I missed this before, but you can unpack the tuple directly here instead of assigning it to `skip_count`

Reply:
Ooooh I really like this, fixing.